File: app-miyou/src/main/java/com/miyou/app/application/credit/usecase/CreditChargeUseCase.java
package com.miyou.app.application.credit.usecase;

import com.miyou.app.domain.credit.model.CreditTransaction;
import com.miyou.app.domain.credit.model.PaymentCharge;
import com.miyou.app.domain.dialogue.model.UserId;
import com.miyou.app.domain.mission.model.MissionId;
import reactor.core.publisher.Mono;

public interface CreditChargeUseCase {

	Mono<CreditTransaction> chargeByPayment(UserId userId, long amount, PaymentCharge source);

	Mono<CreditTransaction> grantSignupBonus(UserId userId);

	Mono<CreditTransaction> grantMissionReward(UserId userId, MissionId missionId, long amount,
		String missionType);

	Mono<Void> initializeIfAbsent(UserId userId);
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/credit/usecase/CreditChargeUseCase.kt
package com.miyou.app.application.credit.usecase

import com.miyou.app.domain.credit.model.CreditTransaction
import com.miyou.app.domain.credit.model.PaymentCharge
import com.miyou.app.domain.mission.model.MissionId
import reactor.core.publisher.Mono

interface CreditChargeUseCase {
    fun chargeByPayment(
        userId: String,
        amount: Long,
        source: PaymentCharge,
    ): Mono<CreditTransaction>

    fun grantSignupBonus(userId: String): Mono<CreditTransaction>

    fun grantMissionReward(
        userId: String,
        missionId: MissionId,
        amount: Long,
        missionType: String,
    ): Mono<CreditTransaction>

    fun initializeIfAbsent(userId: String): Mono<Void>
}
